res/res_pjsip_device_features.c

```diff
@@ -0,0 +1,1005 @@
+{
+	/*** DOCUMENTATION
+		<managerEvent language="en_US" name="DeviceFeatureSync">
+			<managerEventInstance class="EVENT_FLAG_CALL">
+				<synopsis>Raised when a SIP device tries to synchronize a server-side feature</synopsis>
+					<syntax>
+						<parameter name="DeviceID">
+							<para>The unique device ID sent by the endpoint.</para>
+							<para>The same device ID can be used in NOTIFY requests, but this is not used within Broadworks.</para>
+						</parameter>
+						<parameter name="Endpoint">
+							<para>The endpoint (device name).</para>
+						</parameter>
+						<parameter name="CallerIDNum">
+							<para>The Caller ID number of the endpoint.</para>
+						</parameter>
+						<parameter name="Feature">
+							<para>Will be one of the following:</para>
+							<enumlist>
+								<enum name="DoNotDisturb"/>
+								<enum name="CallForwardingAlways"/>
+								<enum name="CallForwardingBusy"/>
+								<enum name="CallForwardingNoAnswer"/>
+							</enumlist>
+						</parameter>
+						<parameter name="Status">
+							<para>1 if enabled or 0 if disabled.</para>
+						</parameter>
+						<parameter name="RingCount">
+							<para>For Call Forwarding No Answer, the ring count.</para>
+						</parameter>
+						<parameter name="Data">
+							<para>For call forwarding enablement, the forward-to number.</para>
+						</parameter>
+					</syntax>
+			</managerEventInstance>
+		</managerEvent>
+	***/
+
+	if (status != 0 && status != 1) {
+		ast_log(LOG_WARNING, "Invalid status: %d\n", status);
+		return -1;
+	}
+
+	ast_debug(3, "%s (%s): %s => %d\n", endpoint, device_id, feature, status);
+
+	manager_event(EVENT_FLAG_CALL, "DeviceFeatureSync",
+		"DeviceID: %s\r\n"
+		"Endpoint: %s\r\n"
+		"CallerIDNum: %s\r\n"
+		"Feature: %s\r\n"
+		"Status: %d\r\n"
+		"RingCount: %d\r\n"
+		"Data: %s\r\n",
+		device_id, endpoint, number, feature, status, ring_count, S_OR(data, ""));
+	return 0;
+}
+
+#define DEVICE_FEATURE_UPDATE(endpoint, db, upd, var, value, dbvalue) { \
+	feature_state_sub->sync_data.var = value; \
+	feature_state_sub->sync_data.upd = 1; \
+	ast_db_put("subscription_device_features/" db, endpoint, dbvalue); \
+}
+
+#define DEVICE_FEATURE_UPDATE_STR(endpoint, db, upd, var, value, dbvalue) { \
+	ast_copy_string(feature_state_sub->sync_data.var, value, sizeof(feature_state_sub->sync_data.var)); \
+	feature_state_sub->sync_data.upd = 1; \
+	ast_db_put("subscription_device_features/" db, endpoint, dbvalue); \
+}
+
+static int parse_incoming_xml(struct feature_state_subscription *feature_state_sub, char *xmlbody, int xmllen, const char *endpoint_name, const char *number)
+{
+	int res = 0;
+	const char *nodename;
+	const char *tmpstr;
+	struct ast_xml_node *root, *tmpnode = NULL;
+	struct ast_xml_doc *xmldoc = ast_xml_read_memory(xmlbody, xmllen);
+
+	if (!xmldoc) {
+		ast_log(LOG_WARNING, "Failed to parse as XML: %.*s\n", xmllen, xmlbody);
```

Comment:
This needs to be more detailed for a user, such as including the endpoint name and what actually happened to cause this to occur